lib/device.c

```diff
@@ -138,6 +138,40 @@ static inline int xrecv(SOCKET s, void *buf, size_t len, int flags)
 #endif
 }
 
+static inline int sockio_poll(struct sync_device *d, int *res_readable, int *res_writeable)
+{
+	assert(d);
+	assert(d->sockio_ctxt);
+	assert(res_readable || res_writeable);
+	return d->sockio_cb.poll(d->sockio_ctxt, res_readable, res_writeable);
+}
+
+static inline int sockio_send(struct sync_device *d, const void *buf, int len)
+{
+	assert(d);
+	assert(d->sockio_cb_ctxt);
+	assert(buf);
+	assert(len > 0);
+	return d->sockio_cb.send(d->sockio_cb_ctxt, buf, len) != len;
+}
+
+static inline int sockio_recv(struct sync_device *d, void *buf, int size)
+{
+	assert(d);
+	assert(d->sockio_cb_ctxt);
+	assert(buf);
+	assert(size > 0);
+	return d->sockio_cb.recv(d->sockio_cb_ctxt, buf, size) != size;
+}
+
+static inline void sockio_close(struct sync_device *d)
+{
+	assert(d);
+	assert(d->sockio_cb_ctxt);
+	d->sockio_cb.close(d->sockio_cb_ctxt);
+	d->sockio_cb_ctxt = NULL;
+}
```

Comment:
I would drop the asserts here. We don't usually assert trivial preconditions (unless they're particularly tricky). And also a lot of these don't really matter to these functions themselves, so better leave that for where it actually matters...